extensions/consensus/payloads.go

```diff
@@ -0,0 +1,26 @@
+package consensus
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/kwilteam/kwil-db/common"
+	"github.com/kwilteam/kwil-db/core/types/transactions"
+)
+
+// Payload defines a new transaction payload, including it's type and the
+// Route (pricer plus execution handler for the TxApp). There does not seem to
+// be a purpose for including the underlying type or a codec for it here since
+// the route is the sole interpreter of the payload.
+type Payload struct {
+	Type  transactions.PayloadType
+	Route Route // must be from here or common, not an internal
+}
+
+// Route is type that TxApp requires to execute externally defined routes.
+type Route interface {
+	Name() string
+	Price(ctx context.Context, app *common.App, tx *transactions.Transaction) (*big.Int, error)
+	PreTx(ctx common.TxContext, app *common.App, tx *transactions.Transaction) (transactions.TxCode, error)
```

Comment:
There was at least one existing route that had some operations to do before creating the inner savepoint.  It did make sense to, say, unmarshal or do some other validation before that, so I went with it despite being very possible to just have the one method, inside the db tx.

Regarding `call`s, these Routes should be exclusive to TxApp and tied to consensus.  Might be missing something, but these Routes (or any TxApp) don't apply to anything done directly from the RPC service.

Reply:
Yeah, looking again at the diff for internal/txapp/routes.go, I think there's a pretty decent case for operations to complete (e.g. unmarshalling, arg validation, etc) before bothering the db backend with a savepoint.

I will definitely have another look at concurrency issue arising from TxApp being used for Price purposes by the RPC or other services that work concurrently with tx execution.

Reply:
Yeah you're totally right, routes are only called during block execution.

Regarding the PreTx and InTx stuff: I think it is important to wrap anything the user does in a tx (even in pre-tx). Something very unintuitive is that if a user returns an error from PreTx, the outer tx will fail to commit the spend. This is why we have to make the second nested transactions in the first place.

Reply:
> Regarding the PreTx and InTx stuff: I think it is important to wrap anything the user does in a tx (even in pre-tx). Something very unintuitive is that if a user encounters a database error from PreTx, the outer tx will fail to commit the spend. This is why we have to make the second nested transactions in the first place. At this point, it then removes the need for a PreTx in the first place.

This is crafted with that in mind.  Have a look at `(*baseRoute) Execute` and where it calls each of the implementation methods.  We have `BeginTx` *twice* there, so two "inner" transactions.  This was an important decision of https://github.com/kwilteam/kwil-db/commit/cb4343db404c589b8d39f91736c097812237cc37, as you point out.

Reply:
The flow is:

- tx1:
   - spend
   - defer Commit tx1
   - PreTx()
   - tx2:
      - InTx()
      - Commit tx2
- return 

If an invalid SQL statement is executed by the user at PreTx, then the function will immediately be returned and commit will be called on tx1. Calling Commit there would fail, since an invalid SQL statement was executed against the tx.

Reply:
Oh yeah, duh, that's for walking through that.  I'll reconsider the PreTx inputs since on reflection it's totally not intended to use the DB at that stage.  Looks like I was being lazy by giving it an entire `common.App` when it really does not use it all all in any PreTx implementation.  At best it would need a logger and possibly one of the read-only methods of Engine that does not execute any queries.  You are right.